com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -1615,7 +1615,12 @@ private IEnumerator ApprovalTimeout(ulong clientId)
 
                 if (IsServer)
                 {
-                    DisconnectClient(clientId);
+                    // Only disconnect when the connection times out.
+                    // If the connection is not approved it will automatically disconnect the client
+                    if (timedOut)
+                    {
+                        DisconnectClient(clientId);
+                    }
```

Comment:
First, I'd like to quote your (@NoelStephensUnity) comment inline here (for the discussion):
> With other transports the removal of client id mapping can happen immediately (server side) when the client is not approved yet the coroutine is left running. Upon timing out (a non-approved server-side client time out coroutine), the server then attempts to disconnect the client which ends up throwing an exception. This doesn't happen with UnityTransport (currently) so this PR is to just to track the coroutines and end them when the client or the server side have received some form of approval response.

Reply:
Also quoting an exception from the linked issue (#2389):

> ```KeyNotFoundException: The given key '1' was not present in the dictionary.
> System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) (at <d6232873609549b8a045fa15811a5bd3>:0)
> Unity.Netcode.NetworkManager.ClientIdToTransportId (System.UInt64 clientId) (at Library/PackageCache/com.unity.netcode.gameobjects@1.2.0/Runtime/Core/NetworkManager.cs:1781)
> Unity.Netcode.NetworkManager.DisconnectRemoteClient (System.UInt64 clientId) (at Library/PackageCache/com.unity.netcode.gameobjects@1.2.0/Runtime/Core/NetworkManager.cs:1361)
> Unity.Netcode.NetworkManager.DisconnectClient (System.UInt64 clientId, System.String reason) (at Library/PackageCache/com.unity.netcode.gameobjects@1.2.0/Runtime/Core/NetworkManager.cs:2063)
> Unity.Netcode.NetworkManager.DisconnectClient (System.UInt64 clientId) (at Library/PackageCache/com.unity.netcode.gameobjects@1.2.0/Runtime/Core/NetworkManager.cs:2038)
> Unity.Netcode.NetworkManager+<ApprovalTimeout>d__180.MoveNext () (at Library/PackageCache/com.unity.netcode.gameobjects@1.2.0/Runtime/Core/NetworkManager.cs:1765)
> UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at <bae255e3e08e46f7bc2fbd23dde96338>:0)```

Reply:
I think we should not change the behavior here — the minimum fix is to simply handle `NM.DisconnectClient()` call gracefully and leave other issues/exceptions related to transport, including 3rd-party ones, up their own implementation.

As a rule of thumb, a transport should be able to handle a disconnect or a send message call to an unknown or "just-removed" connection gracefully.

Perhaps this would give the general idea:

```diff
internal ulong TransportIdToClientId(ulong transportId)
{
-    return transportId == m_ServerTransportId ? ServerClientId : m_TransportIdToClientIdMap[transportId];
+    if (transportId == m_ServerTransportId)
+    {
+        return ServerClientId;
+    }
+
+    if (m_TransportIdToClientIdMap.TryGetValue(transportId, out var clientId))
+    {
+        return clientId;
+    }
+
+    return default;
}

internal ulong ClientIdToTransportId(ulong clientId)
{
-    return clientId == ServerClientId ? m_ServerTransportId : m_ClientIdToTransportIdMap[clientId];
+    if (clientId == ServerClientId)
+    {
+        return m_ServerTransportId;
+    }
+
+    if (m_ClientIdToTransportIdMap.TryGetValue(clientId, out var transportId))
+    {
+        return transportId;
+    }
+
+    return default;
}
```

Reply:
Yeah, not sure why I approached it this way.
Fixing the transport id mapping is indeed a less intrusive fix for the issue.